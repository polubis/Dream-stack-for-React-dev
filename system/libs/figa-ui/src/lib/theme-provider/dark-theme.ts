import type { Theme } from './defs';
import { tokens } from './tokens';

const dark: Theme = {
  outline: {
    color: tokens.common.white,
  },
  switch: {
    bg: tokens.gray[50],
    color: tokens.common.white,
    active: {
      bg: tokens.green[200],
    },
  },
  scroll: {
    thumb: tokens.gray[600],
    track: tokens.gray[550],
    thumbHover: tokens.gray[650],
  },
  divider: {
    default: {
      bg: tokens.common.white,
    },
    primary: {
      bg: tokens.primary[100],
    },
  },
  radio: {
    borderColor: tokens.common.white,
    active: {
      bg: tokens.green[200],
      borderColor: tokens.green[200],
    },
  },
  checkbox: {
    borderColor: tokens.common.white,
    checked: {
      bg: tokens.primary[50],
      borderColor: tokens.primary[50],
      color: tokens.common.white,
    },
  },
  tabs: {
    filled: {
      bg: tokens.gray[500],
      active: {
        borderColor: tokens.green[50],
      },
    },
  },
  chips: {
    color: tokens.common.white,
    borderColor: tokens.common.white,
    active: {
      color: tokens.green[50],
      borderColor: tokens.green[50],
    },
  },
  badge: {
    outlined: {
      primary: {
        color: tokens.primary[50],
        borderColor: tokens.primary[50],
      },
      secondary: {
        color: tokens.primary[100],
        borderColor: tokens.primary[100],
      },
      ok: {
        color: tokens.green[100],
        borderColor: tokens.green[100],
      },
      casual: {
        color: tokens.gray[150],
        borderColor: tokens.gray[150],
      },
    },
    filled: {
      primary: {
        color: tokens.common.black,
        bg: tokens.primary[50],
      },
      secondary: {
        color: tokens.common.black,
        bg: tokens.primary[100],
      },
      ok: {
        color: tokens.common.black,
        bg: tokens.green[100],
      },
      casual: {
        color: tokens.common.black,
        bg: tokens.gray[150],
      },
    },
  },
  thumbnail: {
    bg: tokens.backdrop[100],
    color: tokens.common.white,
  },
  filePicker: {
    outline: {
      color: tokens.primary[50],
    },
    invalid: {
      color: tokens.red[150],
      outline: tokens.red[150],
    },
  },
  field: {
    invalid: {
      color: tokens.red[100],
    },
    hint: {
      color: tokens.common.white,
    },
  },
  font: {
    default: {
      color: tokens.common.white,
    },
    primary: {
      color: tokens.primary[100],
    },
  },
  body: {
    bg: tokens.common.black,
  },
  alert: {
    filled: {
      info: {
        bg: tokens.gray[150],
        color: tokens.common.black,
      },
      ok: {
        bg: tokens.green[150],
        color: tokens.common.black,
      },
      error: {
        bg: tokens.red[50],
        color: tokens.common.white,
      },
      warn: {
        bg: tokens.yellow[50],
        color: tokens.common.black,
      },
    },
    outlined: {
      info: {
        color: tokens.common.white,
      },
      ok: {
        color: tokens.green[150],
      },
      error: {
        color: tokens.red[150],
      },
      warn: {
        color: tokens.yellow[50],
      },
    },
  },
  modal: {
    bg: tokens.common.black,
    border: tokens.dark[50],
    backdrop: tokens.backdrop[50],
  },
  button: {
    filled: {
      primary: {
        color: tokens.common.black,
        bg: tokens.primary[50],
        outlineColor: tokens.common.white,
      },
      secondary: {
        color: tokens.common.black,
        bg: tokens.primary[100],
        outlineColor: tokens.common.white,
      },
      tertiary: {
        color: tokens.common.black,
        bg: tokens.common.white,
        outlineColor: tokens.common.white,
      },
    },
    outlined: {
      primary: {
        color: tokens.primary[50],
        bg: 'transparent',
        borderColor: tokens.primary[50],
        outlineColor: tokens.common.white,
      },
      secondary: {
        color: tokens.primary[100],
        bg: 'transparent',
        borderColor: tokens.primary[100],
        outlineColor: tokens.common.white,
      },
      tertiary: {
        color: tokens.common.white,
        bg: 'transparent',
        borderColor: tokens.common.white,
        outlineColor: tokens.common.white,
      },
    },
    ghost: {
      hoverBg: tokens.dark[250],
    },
  },
  select: {
    bg: tokens.dark[50],
    color: tokens.common.white,
    placeholderColor: tokens.gray[50],
    hoverBg: tokens.gray[100],
    optionSeparator: tokens.gray[100],
    optionActiveColor: tokens.secondary[50],
  },
  nav: {
    bg: tokens.dark[100],
    borderColor: tokens.gray[300],
    bgWithOpacity: tokens.dark[200],
  },
  logo: {
    text: {
      first: tokens.purple[100],
      second: tokens.common.white,
      last: tokens.secondary[200],
    },
    bg: {
      first: {
        from: tokens.red[50],
        to: tokens.primary[250],
      },
      second: {
        from: tokens.secondary[250],
        to: tokens.primary[250],
      },
      last: {
        from: tokens.blue[200],
        to: tokens.primary[250],
      },
    },
  },
  progressCircle: {
    bg: tokens.dark[50],
  },
  input: {
    filled: {
      bg: tokens.dark[50],
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    outlined: {
      borderColor: tokens.common.white,
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    empty: {
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    invalid: tokens.red[150],
  },
  placeholder: {
    filled: {
      bg: tokens.gray[300],
    },
    outlined: {
      bg: tokens.common.black,
      borderColor: tokens.gray[300],
    },
  },
  box: {
    filled: {
      bg: tokens.gray[300],
    },
    outlined: {
      bg: tokens.common.black,
      borderColor: tokens.gray[300],
    },
  },
  footer: {
    bg: tokens.common.black,
    borderColor: tokens.gray[300],
  },
  code: {
    bg: tokens.dark[150],
    gutters: tokens.dark[150],
    selectionBg: '#4d4d4d',
    foldPlaceholder: '#4d4d4d',
    tags: [
      { tag: 'angleBracket', color: '#da6bad' },
      { tag: 'arithmeticOperator', color: '#fff' },
      { tag: 'bitwiseOperator', color: '#fff' },
      { tag: 'blockComment', color: '#6A9955' },
      { tag: 'bool', color: '#569cd6' },
      { tag: 'brace', color: '#ffd70f' },
      { tag: 'bracket', color: '#da6bad' },
      { tag: 'className', color: '#3abfb0' },
      { tag: 'comment', color: '#6A9955' },
      { tag: 'compareOperator', color: '#fff' },
      { tag: 'controlKeyword', color: '#eba0d9' },
      { tag: 'definition', color: '#4fc1ff' },
      { tag: 'definitionKeyword', color: '#3981c5' },
      { tag: 'definitionOperator', color: '#fff' },
      { tag: 'derefOperator', color: '#fff' },
      { tag: 'function', color: '#dcdc9f' },
      { tag: 'keyword', color: '#3981c5' },
      { tag: 'labelName', color: '#4fc1ff' },
      { tag: 'lineComment', color: '#6A9955' },
      { tag: 'literal', color: '#ef8c59' },
      { tag: 'logicOperator', color: '#fff' },
      { tag: 'moduleKeyword', color: '#eba0d9' },
      { tag: 'name', color: '#4fc1ff' },
      { tag: 'null', color: '#569cd6' },
      { tag: 'number', color: '#accea8' },
      { tag: 'operatorKeyword', color: '#3abfb0' },
      { tag: 'paren', color: '#da6bad' },
      { tag: 'typeName', color: '#3abfb0' },
    ],
  },
  codeBlock: {
    header: {
      bg: tokens.gray[350],
      dots: {
        first: {
          bg: tokens.green[50],
        },
        second: {
          bg: tokens.yellow[50],
        },
        third: {
          bg: tokens.blue[650],
        },
      },
    },
    content: {
      bg: tokens.dark[150],
    },
  },
  list: {
    marker: {
      bg: tokens.primary[50],
    },
  },
  blockquote: {
    bg: tokens.gray[450],
    borderColor: tokens.gray[300],
  },
  loader: {
    primary: tokens.common.white,
    secondary: tokens.primary[50],
  },
  avatars: {
    restCounter: {
      bg: tokens.primary[50],
      color: tokens.common.black,
    },
  },
  creatorLayout: {
    borderColor: tokens.gray[300],
    bg: tokens.common.black,
  },
};

export { dark };
