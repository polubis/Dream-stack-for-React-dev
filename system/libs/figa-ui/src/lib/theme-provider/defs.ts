import type { ReactNode } from 'react';

interface ThemeProviderProps {
  children: ReactNode;
}

interface Theme {
  font: {
    color: string;
  };
  background: {
    body: string;
  };
}

interface Tokens {
  [key: string]: {
    [key: string]: string;
  };
}

type ThemeKey = 'dark' | 'light';
type Themes = Record<ThemeKey, Theme>;
type ThemesList = [ThemeKey, Theme][];

interface ThemeProviderValue {
  key: ThemeKey;
  theme: Theme;
  themes: Themes;
  themesList: ThemesList;
  setTheme: (key: ThemeKey) => void;
}

type UnsafeThemeProviderValue = ThemeProviderValue | null;

type GetDefaultValue = (
  value?: Partial<Pick<ThemeProviderValue, 'key' | 'setTheme'>>
) => ThemeProviderValue;

export type {
  ThemeProviderProps,
  Theme,
  ThemeProviderValue,
  UnsafeThemeProviderValue,
  ThemeKey,
  Tokens,
  Themes,
  ThemesList,
  GetDefaultValue,
};
