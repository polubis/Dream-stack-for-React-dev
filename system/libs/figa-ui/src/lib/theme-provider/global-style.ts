import { createGlobalStyle } from 'styled-components';
import { tokens } from './themes';
import {
  appearIn,
  center,
  central,
  column,
  shape,
  font,
  row,
  size,
  streched,
} from '../shared';
import { T_DOWN } from './viewport';

const GlobalStyle = createGlobalStyle`
    @font-face {
        font-family: 'LexendBold';
        src: url('/fonts/LexendBold.ttf') format('truetype');
    }

    @font-face {
        font-family: 'LexendMedium';
        src: url('/fonts/LexendMedium.ttf') format('truetype');
    }
    
    @font-face {
        font-family: 'LexendRegular';
        src: url('/fonts/LexendRegular.ttf') format('truetype');
    }

    @font-face {
        font-family: 'LexendLight';
        src: url('/fonts/LexendLight.ttf') format('truetype');
    }

    html {
        font-size: 62.5%;
        height: 100%;
    }

    body {
        height: 100%;
        background: ${(props) => props.theme.body.bg};
        color: ${(props) => props.theme.font.default.color};
    }

    .sb-main-padded {
        padding: 0 !important;
        margin: 0 !important;
    }

    h1,
    h2,
    h3,
    h4,
    h5,
    h6,
    p,
    figure,
    picture,
    ul,
    ol,
    body,
    html {
        margin: 0;
        padding: 0;
    }

    * {
        box-sizing: border-box;
    }

    /* font.tsx */

    .font {
        &.default, &.default > * {
            color: ${(props) => props.theme.font.default.color};
        }

        &.primary, &.primary > *  {
            color: ${(props) => props.theme.font.primary.color};
        }

        &.h1 {
            ${font('9.8rem', '-1.5px', 'LexendLight', 300)}
        }

        &.h2 {
            ${font('6.1rem', '-0.5px', 'LexendLight', 300)}
        }

        &.h3 {
            ${font('4.9rem', '0px', 'LexendRegular', 400)}
        }

        &.h4 {
            ${font('3.5rem', '0.25px', 'LexendRegular', 400)}
        }

        &.h5 {
            ${font('2.4rem', '0px', 'LexendRegular', 400)}
        }

        &.h6 {
            ${font('2rem', '0.15px', 'LexendMedium', 500)}
        }

        &.b1 {
            ${font('1.6rem', '0.15px', 'LexendRegular', 400)}
            line-height: 2.4rem;
        }

        &.b2 {
            ${font('1.4rem', '0.1px', 'LexendMedium', 500)}
        }

        &.italic, &.italic > * {
            font-style: italic;
        }

        &.bold, &.bold > * {
            font-weight: bold;
            letter-spacing: 0.40px;
        }
    }

    /* font.tsx */

    /* button.tsx */

    .button {
        cursor: pointer;
        border: none;
    }

    .button-size-1 {
        ${font('1.3rem', '0.15px', 'LexendMedium', 500)}
    }

    .button-size-2 {
        ${font('1.5rem', '0.15px', 'LexendMedium', 500)}
    }

    .button-size-3 {
        ${font('1.8rem', '0.15px', 'LexendMedium', 500)}
    }

    .button-size-4 {
        ${font('2.1rem', '0.15px', 'LexendMedium', 500)}
    }

    .button-size-5 {
        ${font('2.4rem', '0.15px', 'LexendMedium', 500)}
    }

    .button-rectangle {
        border-radius: ${tokens.radius[50]};
    }

    .button-rounded {
        ${center()}
        border-radius: ${tokens.radius[1000]};
    }

    .button-filled.button-primary {
        color: ${(props) => props.theme.button.filled.primary.color};
        background: ${(props) => props.theme.button.filled.primary.bg};

        .icon path {
            fill: ${(props) => props.theme.button.filled.primary.color};
        }
    }

    .button-outlined.button-primary {
        color: ${(props) => props.theme.button.outlined.primary.color};
        background: ${(props) => props.theme.button.outlined.primary.bg};
        border: 2px solid ${(props) =>
          props.theme.button.outlined.primary.borderColor};

        .icon path {
            fill: ${(props) => props.theme.button.outlined.primary.color};
        }
    }

    .button-rounded.button-size-1 {
        ${size(tokens.spacing[400])}
    }

    .button-rounded.button-size-2 {
        ${size(tokens.spacing[500])}
    }

    .button-rounded.button-size-3 {
        ${size(tokens.spacing[600])}
    }

    .button-rounded.button-size-4 {
        ${size(tokens.spacing[700])}
    }

    .button-rounded.button-size-5 {
        ${size(tokens.spacing[800])}
    }

    .button-rectangle.button-size-1 {
        ${size(tokens.spacing[400], 'max-content')}
        padding: 0 ${tokens.spacing[100]};
    }

    .button-rectangle.button-size-2 {
        ${size(tokens.spacing[500], 'max-content')}
        padding: 0 ${tokens.spacing[150]};
    }

    .button-rectangle.button-size-3 {
        ${size(tokens.spacing[600], 'max-content')}
        padding: 0 ${tokens.spacing[200]};
    }

    .button-rectangle.button-size-4 {
        ${size(tokens.spacing[700], 'max-content')}
        padding: 0 ${tokens.spacing[250]};
    }

    .button-rectangle.button-size-5 {
        ${size(tokens.spacing[800], 'max-content')}
        padding: 0 ${tokens.spacing[300]};
    }

    /* button.tsx */

    /* modal.tsx */

    .modal {
        ${central('fixed')}
        ${column()}
        min-width: 280px;
        min-height: 280px;
        max-height: 96vh;
        padding: ${tokens.spacing[250]};
        background: ${(props) => props.theme.modal.bg};
        border-radius: ${tokens.radius[50]};
        border: 1px solid ${(props) => props.theme.modal.border};
        z-index: ${tokens.z[500]};
        overflow-y: auto;
    }

    .backdrop {
        ${streched('fixed')}
        background: ${(props) => props.theme.modal.backdrop};
        z-index: ${tokens.z[450]};
    }
    
    /* modal.tsx */

    /* select.tsx */

    .select {
        position: relative;
        width: 100%;

        .select-expander {
            ${size(tokens.spacing[500], '100%')}
            ${font('1.5rem', '0.15px', 'LexendMedium', 500)}
            ${row()}
            padding: 0 ${tokens.spacing[150]};
            border-radius: ${tokens.radius[50]};
            background: ${(props) => props.theme.select.bg};
            color: ${(props) => props.theme.select.color};
            min-width: 220px;
            cursor: pointer;
            border: none;

            &.select-expander-empty {
                color: ${(props) => props.theme.select.placeholderColor};
            }

            &.select-expander-opened {
                background: ${(props) => props.theme.select.hoverBg};
                color: ${(props) => props.theme.select.color};
                cursor: initial;
            }

            &:hover {
                background: ${(props) => props.theme.select.hoverBg};
                color: ${(props) => props.theme.select.color};
            }
        }

        .select-list {
            ${appearIn(tokens.spacing[200], tokens.spacing[100])}
            position: absolute;
            list-style: none;
            margin: 0;
            padding: 0;
            left: 0;
            transform: translateY(8px);
            width: 100%;
            background: ${(props) => props.theme.select.bg};
            border-radius: ${tokens.radius[50]};
            z-index: ${tokens.z[50]};

            .select-list-option {
                ${font('1.4rem', '0.15px', 'LexendMedium', 500)}
                padding: ${tokens.spacing[150]};
                cursor: pointer;

                &:not(:last-of-type) {
                    border-bottom: 1px solid ${(props) =>
                      props.theme.select.optionSeparator};
                }

                &.select-list-option-active {
                    color: ${(props) => props.theme.select.optionActiveColor};
                    cursor: initial;
                }

                &:hover:not(.select-list-option-active) {
                    background: ${(props) => props.theme.select.hoverBg};
                }
            }
        }
    }

    /* select.tsx */

    /* link.tsx */

    .link {
        & > * {
            text-decoration: none;

            &:hover {
                text-decoration: underline;
                text-underline-offset: ${tokens.spacing[100]};
            }
        }
    }

    /* link.tsx */

    /* navigation.tsx */

    .navigation {
        ${row()}
        justify-content: space-between;
        height: 100px;
        border-bottom: 1px solid ${(props) =>
          props.theme.navigation.borderColor};
        background: ${(props) => props.theme.navigation.bg};
        padding: 0 ${tokens.spacing[250]};

        .navigation-links {
            ${row()}
            list-style: none;
            margin: 0 ${tokens.spacing[400]};
            overflow-x: auto;
            max-width: 1000px;
            height: 100%;

            .navigation-link {
                flex-shrink: 0;

                &:not(:last-of-type) {
                    margin-right: ${tokens.spacing[550]};
                }
            }
        }
    }

    /* navigation.tsx */

    /* logo.tsx */

    .logo {
        ${row()}

        .logo-text {
            cursor: pointer;
            flex-flow: wrap;
            margin-left: ${tokens.spacing[200]};

            span {
                ${font('2rem', '0.15px', 'LexendBold', 500)}
            }

            .logo-text-top {
                & > *:first-child {
                    color: ${(props) => props.theme.logo.text.first};
                }

                & > *:last-child {
                    color: ${(props) => props.theme.logo.text.second};
                }
            }

            .logo-text-bottom {
                color: ${(props) => props.theme.logo.text.last};
            }
        }
    }

    /* logo.tsx */

    /* logo-graphic.tsx */

    .logo-graphic {
        cursor: pointer;

        .logo-bg-first-from {
            stop-color: ${(props) => props.theme.logo.bg.first.from};
        }

        .logo-bg-first-to {
            stop-color: ${(props) => props.theme.logo.bg.first.to};
        }

        .logo-bg-second-from {
            stop-color: ${(props) => props.theme.logo.bg.second.from};
        }

        .logo-bg-second-to {
            stop-color: ${(props) => props.theme.logo.bg.second.to};
        }

        .logo-bg-last-from {
            stop-color: ${(props) => props.theme.logo.bg.last.from};
        }

        .logo-bg-last-to {
            stop-color: ${(props) => props.theme.logo.bg.last.to};
        }
    }

    /* logo-graphic.tsx */

    /* emoji-picker.tsx */

    .emoji-picker {
        .emoji-picker-header {
            text-align: center;
            margin-bottom: ${tokens.spacing[300]};
        }

        .emoji-picker-list {
            display: grid;
            grid-template-columns: auto auto auto;
            gap: ${tokens.spacing[200]};
            justify-content: center;
        }
    }

    /* emoji-picker.tsx */

    /* progress-circle.tsx */

    .progress-circle {
        position: relative;

        .progress-circle-back,
        .progress-circle-front {
            ${streched('absolute')}
            border-radius: ${tokens.radius[1000]};
        }

        .progress-circle-front {
            transition: 0.1s transform linear;
            background: ${(props) => props.theme.progressCircle.bg};
        }

        .progress-circle-text {
            ${central('absolute')}
        }
    }

    /* input.tsx  */

    .input {
        border-radius: ${tokens.radius[50]};
        background: ${(props) => props.theme.input.bg};
        min-width: 220px;

        input, input::placeholder {
            ${font('1.5rem', '0.15px', 'LexendMedium', 500)}
        }

        input {
            padding: ${tokens.spacing[150]};
            border: none;
            width: 100%;
            background: none;
            color: ${(props) => props.theme.input.color};

            &::placeholder {
                color: ${(props) => props.theme.input.placeholderColor};
            }
        }
    }

    /* input.tsx */

    /* layout.tsx  */

    .layout {
        ${column()}

        &.full {
            .layout-content {
                display: flex;
                min-height: calc(100vh - 100px);
                max-width: 100vw;

                & > * {
                    width: 100%;
                }
            }
        }

        .layout-content {
            padding: ${tokens.spacing[350]} ${tokens.spacing[250]};
        }
    }

    /* layout.tsx */

    /* box.tsx  */

    .box {
        &.filled, &.outlined {
            border-radius: ${tokens.radius[50]};
        }

        &.filled {
            background: ${(props) => props.theme.box.bg};
        }

        &.outlined {
            border: 1px solid ${(props) => props.theme.box.borderColor};
        }

        &.column {
            ${column()}
        }

        &.center-row {
            ${center()}
        }

        &.center-column {
            ${center('column')}
        }

        &.row {
            ${row()}
        }
    }

    /* box.tsx */

    /* footer.tsx */

    .footer {
        background: ${(props) => props.theme.footer.bg};
        border-top: 1px solid ${(props) => props.theme.footer.borderColor};

        .footer-container {
            ${column()}
            margin: 0 auto;
            max-width: ${tokens.width[50]};
            padding: ${tokens.spacing[300]} 0;

            .footer-content {
                display: grid;
                grid-template-columns: 1fr 1fr 1fr;
                border-bottom: 1px solid ${(props) =>
                  props.theme.footer.borderColor};
                padding-bottom: ${tokens.spacing[200]};
                
                @media ${T_DOWN} {
                    grid-template-columns: auto;
                    grid-template-rows: auto auto auto;
                }

                & > * {
                    padding: ${tokens.spacing[250]};
                }
            }

            .footer-bar {
                ${row()}
                justify-content: space-between;
                padding: ${tokens.spacing[250]} ${tokens.spacing[250]} 0 ${
  tokens.spacing[250]
};

                .footer-bar-left {
                    ${row()}
                }

                .footer-bar-right {
                    ${row()}
                }
            }
        }
    }

    /* footer.tsx */

    /* code-block.tsx */

    .code-block {
        ${column()}

        .code-block-header {
            ${row()}
            justify-content: space-between;
            padding: ${tokens.spacing[200]};
            background: ${(props) => props.theme.codeBlock.header.bg};
            border-top-right-radius: ${tokens.radius[50]};
            border-top-left-radius: ${tokens.radius[50]};

            .code-block-header-dots {
                ${row()}

                & > *:not(:last-child) {
                    margin-right: ${tokens.spacing[150]};
                }

                & > * {
                    ${shape(tokens.spacing[200], tokens.radius[1000])}

                    &:first-of-type {
                        background: ${(props) =>
                          props.theme.codeBlock.header.dots.first.bg};
                    }

                    &:nth-of-type(2) {
                        background: ${(props) =>
                          props.theme.codeBlock.header.dots.second.bg};
                    }

                    &:last-of-type {
                        background: ${(props) =>
                          props.theme.codeBlock.header.dots.third.bg};
                    }
                }
            }
        }

        .code-block-content {
            padding: ${tokens.spacing[200]};
            background: ${(props) => props.theme.codeBlock.content.bg};
            border-bottom-right-radius: ${tokens.radius[50]};
            border-bottom-left-radius: ${tokens.radius[50]};
        }
    }

    /* code-block.tsx */

    /* left-bar.tsx */

    .left-bar {
        ${column()}
        position: fixed;
        bottom: ${tokens.spacing[250]};
        left:${tokens.spacing[250]};
        transition: 0.3s transform ease-in-out;

        & > *:not(:last-child) {
            margin-bottom: ${tokens.spacing[150]};
        }

        &.visible {
            transform: scale(1);
        }

        &.hidden {
            transform: scale(0);
        }
    }

    /* left-bar.tsx */

    /* icon.tsx */

    .icon {
        path {
            fill: ${(props) => props.theme.font.default.color};
        }
        &.alert-error {
            path {
                fill: ${(props) => props.theme.alert.error.color}};
            }
        }
        &.alert-warning {
            path {
                fill: ${(props) => props.theme.alert.warning.color}};
            }
        }
        &.alert-info {
            path {
                fill: ${(props) => props.theme.alert.info.color}};
                rotate: 180deg;
            }
        }
        &.alert-success {
            path {
                fill: ${(props) => props.theme.alert.success.color}};
            }
        }
    }

    /* icon.tsx */

    /* list.tsx */

    .list {
        list-style: none;

        & > *:not(:last-child) {
            margin-bottom: ${tokens.spacing[150]};
        }

        li {
            ${row()}
            ${font('1.6rem', '0.15px', 'LexendBold', 500)}

            &::before {
                margin-right: ${tokens.spacing[150]};
            }
        }
    }
    
    ul.list {
        li {
            &::before {
                content: '';
                ${shape(tokens.spacing[100], tokens.radius[1000])}
                background: ${(props) => props.theme.list.marker.bg};
            }

        }
    }
    
    ol.list {
        counter-reset: numbers;

        li {
            counter-increment: numbers;

            &::before {
                content: counter(numbers) ". ";
                ${font('2rem', '0.15px', 'LexendBold', 500)}
                color: ${(props) => props.theme.list.marker.bg};
            }
        }
    }

    /* list.tsx */

    /* popover.tsx */

    .popover {
        position: relative;

        &-content {
            ${appearIn(tokens.spacing[0], tokens.spacing[150])}
            position: fixed;
            padding: ${tokens.spacing[250]};
        }
    }

    /* popover.tsx */
`;

export { GlobalStyle };
