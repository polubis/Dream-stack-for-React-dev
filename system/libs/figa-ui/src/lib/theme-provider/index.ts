export * from './theme-provider';
export * from './defs';
