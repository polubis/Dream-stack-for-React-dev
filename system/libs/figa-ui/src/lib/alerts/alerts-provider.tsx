import styled from 'styled-components';
import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { Alert } from '../alert';
import { usePortal } from '@system/figa-hooks';
import { tokens } from '../theme-provider';
import type { AlertData, Alerts, AlertsProps, AlertsValue } from './defs';
import { appearIn, center } from '../shared';

const offset = tokens.spacing[200];

const Container = styled.div`
  ${center('column')}
  position: fixed;
  z-index: ${tokens.z[750]};
  width: max-content;
  height: max-content;

  & > *:not(:last-child) {
    margin-bottom: ${tokens.spacing[150]};
  }

  & > * {
    ${appearIn()}
  }

  &.t-l {
    top: ${offset};
    left: ${offset};
  }

  &.t-c {
    top: ${offset};
    left: 0;
    right: 0;
    margin: 0 auto;
  }

  &.t-r {
    top: ${offset};
    right: ${offset};
  }

  &.c-l {
    top: 0;
    bottom: 0;
    left: ${offset};
    margin: auto 0;
  }

  &.c-c {
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
  }

  &.c-r {
    top: 0;
    bottom: 0;
    right: ${offset};
    margin: auto 0;
  }

  &.b-l {
    bottom: ${offset};
    left: ${offset};
  }

  &.b-c {
    margin: 0 auto;
    bottom: ${offset};
    left: 0;
    right: 0;
  }

  &.b-r {
    bottom: ${offset};
    right: ${offset};
  }
`;

const Context = createContext<AlertsValue | null>(null);

const initialAlerts: Alerts = {
  't-l': [],
  't-c': [],
  't-r': [],
  'c-l': [],
  'c-c': [],
  'c-r': [],
  'b-l': [],
  'b-c': [],
  'b-r': [],
};

const AlertsProvider = ({ children }: AlertsProps) => {
  const [alerts, setAlerts] = useState(initialAlerts);
  const timeoutRefs = useRef(new Map<string, NodeJS.Timeout>());

  const hide = useCallback(({ id, position }: AlertData): void => {
    setAlerts((alerts) => ({
      ...alerts,
      [position]: alerts[position].filter((alert) => alert.id !== id),
    }));
  }, []);

  const value = useMemo(
    (): AlertsValue => ({
      show: (payload) => {
        const position = payload.position ?? 't-r';
        const delay = payload.delay ?? 5000;
        const id = new Date().toISOString();
        const alert: AlertData = {
          ...payload,
          position,
          id,
          delay,
        };

        timeoutRefs.current.set(
          id,
          setTimeout(() => {
            hide(alert);
          }, delay)
        );

        setAlerts((alerts) => ({
          ...alerts,
          [position]: [...alerts[position], alert],
        }));

        return alert;
      },
    }),
    [hide]
  );

  const { render } = usePortal();

  useEffect(() => {
    const timeouts = timeoutRefs.current;

    return () => {
      timeouts.forEach((timeout) => {
        clearTimeout(timeout);
      });
    };
  }, []);

  return (
    <Context.Provider value={value}>
      {children}
      {Object.entries(alerts)
        .filter(([, alertsArr]) => alertsArr.length > 0)
        .map(([position, alertsArr]) =>
          render(
            <Container className={position}>
              {alertsArr.map((alert) => (
                <Alert
                  key={alert.id}
                  maxWidth="420px"
                  {...alert}
                  onClose={() => {
                    hide(alert);
                    alert.onClose?.();
                  }}
                />
              ))}
            </Container>
          )
        )}
    </Context.Provider>
  );
};

const useAlert = () => {
  const ctx = useContext(Context);

  if (!ctx)
    throw Error(
      'Lack of <AlertsProvider> component wrapper in your components tree'
    );

  return ctx;
};

export { AlertsProvider, useAlert };
