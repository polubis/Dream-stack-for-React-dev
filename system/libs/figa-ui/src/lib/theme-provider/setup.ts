import type { Themes } from './defs';
import { light } from './light-theme';
import { dark } from './dark-theme';
import { tokens } from './tokens';

// Check for reference: https://mui.com/joy-ui/customization/theme-colors/

const themes: Themes = {
  light,
  dark,
};

export { themes, tokens };
