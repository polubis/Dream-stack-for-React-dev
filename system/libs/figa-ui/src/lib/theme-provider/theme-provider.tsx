import { useState, createContext, useMemo, useContext } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import type {
  ThemeKey,
  ThemeProviderProps,
  ThemeProviderValue,
  ThemesList,
  UnsafeThemeProviderValue,
  GetDefaultValue,
} from './defs';
import { themes } from './themes';
import { GlobalStyle } from './global-style';

const Ctx = createContext<UnsafeThemeProviderValue>(null);

const defaultKey = 'dark' as ThemeKey;

const getDefaultValue: GetDefaultValue = ({
  key = defaultKey,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  setTheme = () => {},
} = {}) => ({
  key,
  setTheme,
  themes,
  theme: themes[key],
  themesList: Object.entries(themes) as ThemesList,
});

const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [key, setKey] = useState(defaultKey);

  const value: ThemeProviderValue = useMemo(
    () =>
      getDefaultValue({
        key,
        setTheme: setKey,
      }),
    [key]
  );

  return (
    <Ctx.Provider value={value}>
      <StyledThemeProvider theme={value.theme}>
        <GlobalStyle />
        {children}
      </StyledThemeProvider>
    </Ctx.Provider>
  );
};

const useThemeProvider = (): ThemeProviderValue => {
  const ctx = useContext(Ctx);

  return ctx ? ctx : getDefaultValue();
};

export { ThemeProvider, useThemeProvider };
