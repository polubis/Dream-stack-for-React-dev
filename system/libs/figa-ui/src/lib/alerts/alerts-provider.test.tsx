import { act, fireEvent, render, screen } from '@testing-library/react';
import { AlertsProvider, useAlert } from './alerts-provider';
import { ThemeProvider } from '../theme-provider';

jest.useFakeTimers();

describe('Alerts works when: ', () => {
  const ComponentFixture = () => {
    const alert = useAlert();

    return (
      <button
        onClick={() =>
          alert.show({
            children: 'Alert content',
            className: 'my-class',
            delay: 4000,
            variant: 'outlined',
            type: 'ok',
          })
        }
      >
        Show
      </button>
    );
  };

  it('alerts may be opened/closed and are closed after timeouts', async () => {
    const errorSpy = jest.spyOn(console, 'error').mockImplementation(jest.fn());

    render(
      <ThemeProvider>
        <AlertsProvider>
          <ComponentFixture />
        </AlertsProvider>
      </ThemeProvider>
    );

    fireEvent.click(screen.getByText(/Show/));

    screen.getByText(/Alert content/);

    fireEvent.click(screen.getByText(/Show/));
    fireEvent.click(screen.getByText(/Show/));

    expect(screen.getAllByText(/Alert content/)).toHaveLength(3);
    expect(screen.getAllByText(/Alert content/)).toMatchSnapshot();

    await act(() => {
      jest.advanceTimersByTime(5500);
    });

    expect(screen.queryAllByText(/Alert content/)).toHaveLength(0);

    errorSpy.mockRestore();
  });
});
