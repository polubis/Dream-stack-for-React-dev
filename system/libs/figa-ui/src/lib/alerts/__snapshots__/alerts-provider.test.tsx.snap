// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alerts works when:  alerts may be opened/closed and are closed after timeouts 1`] = `
Array [
  <p
    class="font b2 default trim"
  >
    Alert content
  </p>,
  <p
    class="font b2 default trim"
  >
    Alert content
  </p>,
  <p
    class="font b2 default trim"
  >
    Alert content
  </p>,
]
`;
